espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/EspressoW3CActionAdapter.java

```diff
@@ -0,0 +1,37 @@
+package io.appium.espressoserver.lib.helpers.w3c.adapter;
+
+import android.support.test.espresso.InjectEventSecurityException;
+import android.support.test.espresso.UiController;
+
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
+import io.appium.espressoserver.lib.helpers.w3c.dispatcher.KeyDispatch.KeyEvent;
+
+import static android.view.KeyEvent.ACTION_DOWN;
+
+public class EspressoW3CActionAdapter extends BaseW3CActionAdapter {
+
+    private final UiController uiController;
+
+    public EspressoW3CActionAdapter(UiController uiController) {
+        this.uiController = uiController;
+    }
+
+    public boolean keyDown(KeyEvent keyEvent) throws AppiumException {
+        android.view.KeyEvent androidKeyEvent = new android.view.KeyEvent(
+                System.currentTimeMillis(), System.currentTimeMillis(),
+                ACTION_DOWN, Character.getNumericValue(keyEvent.getKey().charAt(0)),
```

Comment:
Character.getNumericValue is not going to work, since android key codes are different from ASCII codes

Reply:
Right, it needs to be one of these: https://android.googlesource.com/platform/frameworks/native/+/master/include/android/keycodes.h

I'll probably just need to tweak it so that you can pass the raw keycode int in the payload instead of a single unicode character.

Reply:
It is not simple at all.

You can also check how I do key codes translation in UIA2 server: https://github.com/appium/appium-uiautomator2-server/blob/cc6fcc6ad7b58e4550e7d09fd5a334d8aef4e8a9/app/src/main/java/io/appium/uiautomator2/handler/W3CActions.java#L142 

I've tested this implementation on Selenium's-java W3C Actions.sendKeys and it worked properly, at least for ASCII characters. BTW, Selenium API itself does not implement the standard fully on the client side. For instance, isAlt/isCtrl and other similar properties are ignored.

Reply:
Okay, yeah I looked it over and I think it makes sense to me now. I should be able to port over those translations. I'll do that once the spec-work is completed and I get to the Espresso implementation.

Implementing the whole standard shouldn't be an issue, if the Espresso adapter doesn't need access to isAlt/isCtrl it's fine to just ignore it.